src/ore/src/id_gen.rs

```diff
@@ -57,6 +58,37 @@ impl<Id: From<u64>> Gen<Id> {
 /// A generator of u64-bit IDs.
 pub type IdGen = Gen<u64>;
 
+/// Manages the allocation of unique IDs.
+///
+/// Atomic version of `Gen`, for sharing between threads.
+#[derive(Debug)]
+pub struct AtomicGen<Id> {
+    id: AtomicU64,
+    phantom: PhantomData<Id>,
+}
+
+impl<Id> Default for AtomicGen<Id> {
+    fn default() -> Self {
+        Self {
+            id: AtomicU64::new(0),
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<Id: From<u64> + Default> AtomicGen<Id> {
+    /// Allocates a new identifier of type `Id` and advances the generator.
+    pub fn allocate_id(&self) -> Id {
+        let id = self.id.fetch_add(1, Ordering::Relaxed);
```

Comment:
Might be worth leaving a comment explaining why `Ordering::Relaxed` is correct (I haven't thought about it myself tbh).

Reply:
I'm also not 100% sure. The docs say "In its weakest Ordering::Relaxed, only the memory directly touched by the operation is synchronized." I _think_ this is sufficient here because all we need is that each user of the atomic gets back a different value, the atomic doesn't protect any other state we require to be synchronized.

This random SO post supports my reasoning: https://stackoverflow.com/questions/30407121/which-stdsyncatomicordering-to-use#33293463

> Relaxed Ordering
> There are no constraints besides any modification to the memory location being atomic (so it either happens completely or not at all). This is fine for something like a counter if the values retrieved by/set by individual threads don't matter as long as they're atomic.

I'll add a comment to that effect. But lmk if you still have doubts! I think it would also be fine to just use the strongest ordering and be done with it.

Reply:
The best resource for understanding these is chapter 3 from Mara Bos' book on atomics and locks. Here is a link to the section for this specific question but I highly recommend reading the whole chapter https://marabos.nl/atomics/memory-ordering.html#relaxed

`Relaxed` sounds right for me too

Reply:
Thanks! I have also enjoyed Herb Sutter's `atomic<> weapons` talk: https://www.youtube.com/watch?v=A8eCGOqgvH4